File: src/simulations/matlab/Fig2/temp_pgd_attack.m
% temp_pgd_attack.m
% Native MATLAB implementation of PGD adversarial attack on the CSTR
% tracking control system from Fig2a.m.
%
% Implements the equivalent of:
%   pgd   = ProjectedGradientDescent(classifier, eps=eps_val, eps_step=tau, max_iter=Npgd)
%   y_adv = pgd.generate(y)
%
% Attack target : temperature measurement y (x2 in the paper)
% Classifier    : sliding-mode sign controller  S = sign(y - r2)
% Loss          : squared tracking error  L(delta) = ||y + delta - r2||^2
% Gradient proxy: surrogate sub-gradient  nabla_delta L = y + delta - r2
% PGD update    : delta = clip(delta + tau*sign(nabla_delta L),  -eps, eps)

%% ── System Parameters (from Fig2a.m) ─────────────────────────────────────────
alpha  = 1;
betta  = 100;
lambda = 0.3;
gamma  = 20;
B      = 1;
Da     = 0.072;
n      = 45000;
tmax   = 45;
dt     = tmax / n;

%% ── PGD Hyperparameters ───────────────────────────────────────────────────────
eps_val = 0.1;   % L-inf ball radius  (epsilon)
tau     = 0.01;  % step size          (alpha / eps_step)
Npgd    = 40;    % iteration count    (max_iter)

%% ── Step 1: Simulate clean system (Fig2a.m) ──────────────────────────────────
x  = zeros(1, n);
y  = zeros(1, n);
t  = zeros(1, n);
r1 = zeros(1, n);
r2 = zeros(1, n);
S  = zeros(1, n);
u  = zeros(1, n);
e1 = zeros(1, n);
e2 = zeros(1, n);

for i = 1:n-1
    t(i+1) = t(i) + dt;

    if     0  <= t(i+1) && t(i+1) <= 15,   r1(i+1) = 0.4472;
    elseif 15 <  t(i+1) && t(i+1) <  30,   r1(i+1) = 0.7646;
    else,                                    r1(i+1) = 0.4472;
    end

    if     0  <= t(i+1) && t(i+1) <= 15,   r2(i+1) = 2.752;
    elseif 15 <  t(i+1) && t(i+1) <  30,   r2(i+1) = 4.7052;
    else,                                    r2(i+1) = 2.752;
    end

    f1 = x(i) + dt*(-alpha*x(i) + Da*(1-x(i))*exp(y(i)/(1+y(i)/gamma)));
    f2 = y(i) + dt*(-alpha*y(i) + B*Da*(1-x(i))*exp(y(i)/(1+y(i)/gamma)));

    if     y(i)-r2(i) > 0,  S(i) =  1;
    elseif y(i)-r2(i) < 0,  S(i) = -1;
    else,                    S(i) =  0;
    end

    u(i) = -1/lambda * (-betta*S(i) + r2(i+1) - r2(i) - f2);

    x(i+1) = f1;
    y(i+1) = y(i) + dt*(-betta*S(i) + r2(i+1) - r2(i));

    e1(i+1) = x(i+1) - r1(i+1);
    e2(i+1) = y(i+1) - r2(i+1);
end

%% ── Step 2: PGD attack ────────────────────────────────────────────────────────
% delta(k+1) = clip( delta(k) + tau * sign( y + delta(k) - r2 ),  -eps, eps )
%
% Rationale: loss L = ||y + delta - r2||^2
%            nabla_delta L = 2*(y + delta - r2)
%            sign(nabla_delta L) = sign(y + delta - r2)
%            We ascend the loss gradient to maximise tracking error.
%
% The gradient direction changes each iteration as delta shifts
% the measurement closer to (or past) the sign-change boundary r2.

delta = zeros(1, n);

for k = 1:Npgd
    nabla = y + delta - r2;            % surrogate gradient  (n x 1 vector op)
    delta = delta + tau * sign(nabla); % gradient ascent step
    delta = max(-eps_val, min(eps_val, delta));  % project onto L-inf ball
end

y_adv = y + delta;                     % adversarial temperature measurements

%% ── Step 3: Simulate attacked system ─────────────────────────────────────────
% Controller sees y_adv; actual plant state evolves from true dynamics.
x_att  = zeros(1, n);
y_att  = zeros(1, n);
S_att  = zeros(1, n);
u_att  = zeros(1, n);
e1_att = zeros(1, n);
e2_att = zeros(1, n);

for i = 1:n-1
    f1_att = x_att(i) + dt*(-alpha*x_att(i) + Da*(1-x_att(i))*exp(y_att(i)/(1+y_att(i)/gamma)));
    f2_att = y_att(i) + dt*(-alpha*y_att(i) + B*Da*(1-x_att(i))*exp(y_att(i)/(1+y_att(i)/gamma)));

    % Controller uses adversarial measurement y_adv instead of y_att
    if     y_adv(i)-r2(i) > 0,  S_att(i) =  1;
    elseif y_adv(i)-r2(i) < 0,  S_att(i) = -1;
    else,                         S_att(i) =  0;
    end

    u_att(i) = -1/lambda * (-betta*S_att(i) + r2(i+1) - r2(i) - f2_att);

    x_att(i+1) = f1_att;
    y_att(i+1) = y_att(i) + dt*(-betta*S_att(i) + r2(i+1) - r2(i));

    e1_att(i+1) = x_att(i+1) - r1(i+1);
    e2_att(i+1) = y_att(i+1) - r2(i+1);
end

%% ── Step 4: Report ───────────────────────────────────────────────────────────
flip_rate = mean(sign(y_adv - r2) ~= S);
fprintf('PGD results  (eps=%.3f  tau=%.4f  Npgd=%d)\n', eps_val, tau, Npgd);
fprintf('  Sign-flip rate     : %.1f%%\n',  flip_rate * 100);
fprintf('  Max  |delta|       : %.4f\n',    max(abs(delta)));
fprintf('  Mean |delta|       : %.4f\n',    mean(abs(delta)));
fprintf('  RMS tracking error clean    e2 : %.4f\n', rms(e2));
fprintf('  RMS tracking error attacked e2 : %.4f\n', rms(e2_att));

%% ── Step 5: Plot ─────────────────────────────────────────────────────────────
figure('Name', 'PGD Attack on CSTR Tracking Control');

subplot(3,1,1)
plot(t, y, 'k-', t, y_att, 'b--', t, r2, 'r-')
ylabel('x_2 (Temperature)')
xlabel('Time (s)')
legend('Clean y', 'Attacked y', 'r_2', 'Location', 'best')
title(sprintf('PGD: \\epsilon=%.2f  \\tau=%.3f  N_{pgd}=%d', eps_val, tau, Npgd))
axis([0 tmax 0 5])
grid on

subplot(3,1,2)
plot(t, e2, 'k-', t, e2_att, 'b--')
yline(0, 'r--')
ylabel('e_2 = y - r_2')
xlabel('Time (s)')
legend('Clean error', 'Attacked error', 'Location', 'best')
axis([0 tmax -3 2.5])
grid on

subplot(3,1,3)
plot(t, delta, 'g-')
ylabel('\delta (Perturbation)')
xlabel('Time (s)')
title(sprintf('Max |\\delta| = %.4f  (L-\\infty bound = %.2f)', max(abs(delta)), eps_val))
axis([0 tmax -eps_val*1.1 eps_val*1.1])
grid on

File: src/simulations/matlab/Fig2/temp_pgd_attack.py
"""
Adversarial PGD attack on CSTR tracking control system.

Translates Fig2a.m to Python, trains a neural network approximation of the
sliding-mode controller, then applies ART's ProjectedGradientDescent to
generate adversarial temperature measurements y_adv.

System: CSTR (Continuous Stirred Tank Reactor)
  x1 = conversion (x in MATLAB)
  x2 = temperature (y in MATLAB)
  Controller: signum sliding-mode, S = sign(x2 - r2)

Attack framing:
  classifier input  : [x2_measured, r2_ref]  (shape: [N, 2])
  classifier output : S class ∈ {-1→0, 0→1, 1→2}  (3 classes)
  pgd.generate(y)   : y = measurement array (N, 2), returns y_adv perturbed
"""

import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from art.attacks.evasion import ProjectedGradientDescent
from art.estimators.classification import PyTorchClassifier

# ── Parameters from Fig2a.m ───────────────────────────────────────────────────
alpha  = 1.0
betta  = 100.0
gamma  = 20.0
B      = 1.0
Da     = 0.072
n      = 45_000
tmax   = 45.0
dt     = tmax / n

# ── Reference signals ─────────────────────────────────────────────────────────
def r1_ref(t: float) -> float:
    if t <= 15:   return 0.4472
    elif t < 30:  return 0.7646
    else:         return 0.4472

def r2_ref(t: float) -> float:
    if t <= 15:   return 2.752
    elif t < 30:  return 4.7052
    else:         return 2.752

# ── Simulate Fig2a.m (Euler integration) ─────────────────────────────────────
def simulate():
    x  = np.zeros(n)
    y  = np.zeros(n)
    t  = np.zeros(n)
    r1 = np.zeros(n)
    r2 = np.zeros(n)

    r1[0] = 0.0
    r2[0] = 0.0

    for i in range(n - 1):
        t[i + 1]  = t[i] + dt
        r1[i + 1] = r1_ref(t[i + 1])
        r2[i + 1] = r2_ref(t[i + 1])

        rxn = Da * (1.0 - x[i]) * np.exp(y[i] / (1.0 + y[i] / gamma))

        e2 = y[i] - r2[i]
        S  = 1.0 if e2 > 0 else (-1.0 if e2 < 0 else 0.0)

        x[i + 1] = x[i] + dt * (-alpha * x[i] + rxn)
        y[i + 1] = y[i] + dt * (-betta * S + r2[i + 1] - r2[i])

    return t, x, y, r1, r2

print("Running Fig2a.m simulation...")
t_traj, x_traj, y_traj, r1_traj, r2_traj = simulate()

# ── Build training dataset ────────────────────────────────────────────────────
# Input:  [x2_measurement, r2_reference]
# Label:  sign class of (x2 - r2)  mapped to {-1→0, 0→1, 1→2}
e2_train = y_traj[:-1] - r2_traj[:-1]
labels   = np.where(e2_train > 0, 2, np.where(e2_train < 0, 0, 1)).astype(np.int64)

X_train = np.stack([y_traj[:-1], r2_traj[:-1]], axis=1).astype(np.float32)

# ── Neural network controller approximation ───────────────────────────────────
class ControllerNet(nn.Module):
    def __init__(self):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(2, 64), nn.Tanh(),
            nn.Linear(64, 64), nn.Tanh(),
            nn.Linear(64, 3),
        )

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.net(x)

model     = ControllerNet()
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=1e-3)

dataset = torch.utils.data.TensorDataset(
    torch.tensor(X_train),
    torch.tensor(labels),
)
loader = torch.utils.data.DataLoader(dataset, batch_size=1024, shuffle=True)

print("Training controller approximation...")
for epoch in range(10):
    total_loss = 0.0
    for xb, yb in loader:
        optimizer.zero_grad()
        loss = criterion(model(xb), yb)
        loss.backward()
        optimizer.step()
        total_loss += loss.item()
    print(f"  epoch {epoch + 1:2d}  loss={total_loss / len(loader):.4f}")

# ── ART classifier wrapper ────────────────────────────────────────────────────
classifier = PyTorchClassifier(
    model=model,
    loss=criterion,
    optimizer=optimizer,
    input_shape=(2,),
    nb_classes=3,
    clip_values=(0.0, 6.0),   # x2 (temperature) physical bounds
)

# ── PGD attack ────────────────────────────────────────────────────────────────
ε    = 0.1    # max perturbation magnitude on temperature measurement
α    = 0.01   # PGD step size
Npgd = 40     # number of PGD iterations

pgd   = ProjectedGradientDescent(classifier, eps=ε, eps_step=α, max_iter=Npgd)

y = X_train[:2000]           # measurement window to attack, shape (N, 2)
y_adv = pgd.generate(y)      # adversarially perturbed measurements

# ── Report ────────────────────────────────────────────────────────────────────
preds_clean = np.argmax(classifier.predict(y),     axis=1)
preds_adv   = np.argmax(classifier.predict(y_adv), axis=1)

flip_rate   = np.mean(preds_clean != preds_adv)
max_delta   = np.max(np.abs(y_adv - y))
mean_delta  = np.mean(np.abs(y_adv - y))

print(f"\nPGD results (ε={ε}, α={α}, Npgd={Npgd}):")
print(f"  Sign-class flip rate : {flip_rate * 100:.1f}%")
print(f"  Max  |δy|            : {max_delta:.4f}")
print(f"  Mean |δy|            : {mean_delta:.4f}")
print(f"\nSample comparison:")
print(f"  y[0]     = {y[0]}")
print(f"  y_adv[0] = {y_adv[0]}")
print(f"  Δ        = {y_adv[0] - y[0]}")
